### PR TITLE
Added cache to CircleCi config

### PR DESCRIPTION
**Pull request by umbernhard**
**Description**
Adds dependency caching and parallelism to CircleCI to speed up testing. 

**Testing**
N/A

**Progress**
Ready to go.

### CODE REVIEW

**Comment by umbernhard** [comment]
@jonahkagan this is ready for review. I'll squash merge it once it's approved. Speeds up testing by about a minute, and that will increase the more tests we add.

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:32]: is it possible to configure this at a higher level in the file, or does it have to be set on each command?

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:5]: what does this change do?

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:30]: is it possible to name these steps so that we can tell which one is which? this is what I see now: 
Screen Shot 2020-09-08 at 2 44 19 PM

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:51]: same question about naming as above

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:49]: do we need this line still?

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:69]: "lint"

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:84]: what does this do? also we lost test coverage I believe, which is actually one of the more important things to run on CI cuz ppl (me) don't often run it locally

File: Makefile
Comment (jonahkagan) [inline Makefile:38]: ooo exciting. pylint is so slow!

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:32]: My understanding is that if you use `parallelism > 1` for sequential steps, it runs those steps that many times. So yes, we could but from my testing that made it slower, as random stuff would cause one run to be way slower than the others.

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:5]: nothing, it's a legacy of my moving stuff into workflows. Will fix :smile:

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:30]: I'll have to check.

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:51]: same answer about naming as above :stuck_out_tongue:

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:49]: Will remove.

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:84]: CircleCI requires some special juju to actually run tests in parallel, you basically have to let it handle the chunking of tests to different nodes. 

Re: coverage I will have to see what I can do about that.

**Comment by umbernhard** [comment]
@jonahkagan okay I think this is ready to go.

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:100]: `make test-server-coverage`?

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:22]: maybe pull out the different caching steps as reusable commands? https://circleci.com/docs/2.0/reusing-config/

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:15]: how did you choose resource_class? do all the jobs benefit from this class? also do they cost diff amounts of money or is it all the same?

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:154]: do we actually need chrome for the frontend tests?

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:247]: surprised that we don't want `make install-development` here...

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:255]: do we need to install chrome separately for the cypress tests (given that the docker image has `browsers` in the name)?

File: client/package.json
Comment (jonahkagan) [inline client/package.json:11]: this command isn't getting used anymore as far as I can tell

File: client/run-cypress-tests.sh
Comment (jonahkagan) [inline client/run-cypress-tests.sh:6]: im pretty sure the `kill 0` was there to kill all the child processes as well. have you tried running this locally? does it leave zombie processes?

File: .circleci/config.yml
Comment (jonahkagan) [inline .circleci/config.yml:126]: why don't we use yarn for everything? there's a `yarn.lock` file in `arlo` and in `arlo/client`

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:100]: That runs it in parallel, which I don't want to do because it breaks stuff.

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:15]: from what I can tell the larger the resource class the faster the tests run. We can probably knock down some of the already fast steps down. Going from medium -> xlarge saves us about a minute of compute time. And it uses more credits, but I don't think we're coming close to our alotted limit that would require us to pay more (except when I'm testing these configs :sweat_smile:)

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:154]: I can check, but I tried removing it and it broke stuff. The docker we're loading actually has browsers in it, but yarn complained about their version so the manual install seems necessary.

File: .circleci/config.yml
Comment (jonahkagan) [reply .circleci/config.yml:100]: oh really? that's kind of concerning. i run them in parallel locally. are we throwing out that parallelism?

File: .circleci/config.yml
Comment (jonahkagan) [reply .circleci/config.yml:15]: cool, sounds fine then

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:22]: Done

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml:188]: @jonahkagan let me know if you have thoughts on this.

File: client/package.json
Comment (umbernhard) [reply client/package.json:11]: isn't it called when `yarn --cwd client test` is called? (see the line right above it)

File: .circleci/config.yml
Comment (umbernhard) [reply .circleci/config.yml]: Fixed this.

File: .circleci/config.yml
Comment (jonahkagan) [reply .circleci/config.yml:188]: i thought it would at least need to install cypress, which is a dev dependency

File: client/package.json
Comment (jonahkagan) [reply client/package.json:11]: oh lol true 🤦

File: client/run-cypress-tests.sh
Comment (jonahkagan) [reply client/run-cypress-tests.sh:6]: that's weird, because i dont think it did that before, did it? i.e. cypress tests have been passing on CI

File: .circleci/config.yml
Comment (jonahkagan) [reply .circleci/config.yml:22]: WOW it looks so good with this change

**Review by jonahkagan — APPROVED** [review]
i think we still have a few outstanding conversations (e.g. the cypress thing) but this is looking amazing. approving pending we resolve those last bits

**Comment by umbernhard** [comment]
@jonahkagan for some reason the cypress script works fine now, so I restored it to the version on main. Otherwise I think we're good to go here, but happy to answer other questions.

**Review by jonahkagan — APPROVED** [review]
SO READY FOR THIS

**Pull request merged**
